### DISCUSSION TITLE
how can i swmm parameter calibration using pyswmm and optimization algorithm?

### DISCUSSION THREAD

**Discussion by boram1024** [pyswmm/pyswmm#504] [Q&A] [answered]
As the title says, how can i swmm parameter calibration using pyswmm and optimization algorithm(ex. GA, DE etc....)?
Has anyone ever done something like this?

**Comment by bemcdonnell** [comment]
@boram1024, I can think of at least 10 researchers and industry people that have put Pyswmm into their optimization workflow.  It’s very useful. Can you articulate some of the challenges that you are having?

**Reply by boram1024** [reply]
First of all, thank you for showing interest in my question.
To be honest, I am a beginner in Python.
Until now, I have been handling parameter optimization problems in SWMM using the R environment(using SWMMR packages).
Maybe because I'm a beginner in Python, I found it difficult to handle Pyswmm in the Python environment.
The most challenging part I encountered initially was setting up the optimization target parameters.
Do you have any examples related to parameter optimization?

**Reply by bemcdonnell** [reply]
@boram1024 could you clarify what sort of parameters you're looking to optimize? Furthermore, what are you objective functions you're building?  If we can get a better understanding of the problem you are trying to solve (and the size of the model) it would help frame the answer.  It's fair to say that PySWMM has been used in many optimization frameworks.  If you are trying to solve a problem that isn't very complex (from a standpoint of optimization), i might suggest something quick and painless.  If you model is huge and will take a long time to run, we might think of parallelization..  Can you frame the problem a little bit? 

As for training, we have an entire YouTube Channel dedicated to training content. This can help you get more comfortable with Python and using PySWMM.   Also, are you a student or are you working in industry?

**Reply by boram1024** [reply]
First of all, I am currently working at a research institution.
The types of parameters can include impermeable area ratio(%Imperv), watershed width(Width), and roughness.
Additionally, the objective function is planned to be composed of RMSE or NSE.
Of course, I have also reviewed some tutorials available on YouTube.
parameter optimization problems can be solved through YouTube tutorials only?

**Reply by bemcdonnell** [reply, marked as answer]
@boram1024 so let’s break apart the problem into steps:

1. design objective (multi objective) function.  Sounds like you have two in mind.  You’re comparing your timeseries model data to measured data.  Like anything else, you’ll need to devise a strategy to sample common time steps to compare. Likely your model and measured data will not have 1-for-1 samples. So you might use python Pandas to resample and interpolate where measured data are missing.  Once you do that, RMSE / NSE is a very straightforward calculation. 
2. next we need to come up with a strategy to edit parameters inside the INP file.  You can use SWMM-api, swmmio, or parse the INP directly and rewrite a new one to change the parameters you want to change. 
3. once that is done, you’ll either pick a standard optimizer like Scipy.optimize or something else and put your model “run” in a function to hand to your optimizer.  Your function will embed the following steps: 1. Open original model, 2. Edit parameters and save new INP file, 3. Run model, 4. Read OUT file to get modeled data for your objective, 5. Overlay/resample both your model and measured data 6. Compute RMSE/NSE, and 7. Return the calculated objective value.  

Things to consider - if your model runs slow, the workflow architecture might require running parallel simulations. This changes things quite a bit. 

Feel free to email me directly. If you can agree to contribute workflow documentation back to the project, I will help you.

**Reply by boram1024** [reply]
2. next we need to come up with a strategy to edit parameters inside the INP file. You can use SWMM-api, swmmio, or parse the INP directly and rewrite a new one to change the parameters you want to change.

==> I thought I had to use pyswmm for this part. Like your comment, I will try to parameter  edit using swmm-api and swmmio.
        Actually, this part was the most challenging for me. 
        
Once again, thank you for your interest.

**Reply by nkarki07** [reply]
Hi, @boram1024. Which optimization algorithm did you use in your case? Any chances you were working with genetic algorithms? I am doing my thesis using that and would love some insights from you. Thank you!

**Reply by bemcdonnell** [reply]
I see it has been quite a while since this question has been raised. Worth noting that with PySWMM-v2 you can now change any model parameter using PySWMM SimPreConfig.  No longer will you need to go to an external dependency

https://pyswmm.github.io/pyswmm/reference/api/pyswmm.simulation.SimulationPreConfig.html

**Reply by nkarki07** [reply]
Thank you @bemcdonnell for the info!
